src/lib/tls/tls_signature_scheme.cpp

```diff
@@ -256,6 +258,76 @@ AlgorithmIdentifier Signature_Scheme::algorithm_identifier() const noexcept
+   //    digest algorithm. [...]  When used in certificate signatures, the
+   //    algorithm parameters MUST be DER encoded.
+   auto pss_params = [](const std::string hash_name)
+      {
+         const AlgorithmIdentifier hash_id(hash_name, AlgorithmIdentifier::USE_NULL_PARAM);
+         const AlgorithmIdentifier mgf_id("MGF1", hash_id.BER_encode());
+
+         std::vector<uint8_t> parameters;
+         DER_Encoder(parameters)
+            .start_sequence()
+            .start_context_specific(0).encode(hash_id).end_cons()
+            .start_context_specific(1).encode(mgf_id).end_cons()
+            .start_context_specific(2).encode(HashFunction::create(hash_name)->output_length()).end_cons()
+            .start_context_specific(3).encode(size_t(1)).end_cons() // trailer field
+            .end_cons();
+
+         return parameters;
+      };
+
+   switch(m_code)
+      {
+      // for RSA PKCSv1.5 parameters "SHALL" be NULL
+      case RSA_PKCS1_SHA1:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-160)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA256:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-256)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA384:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-384)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA512:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-512)", AlgorithmIdentifier::USE_NULL_PARAM);
+
+      // for DSA, ECDSA, GOST parameters "SHALL" be empty
+      case ECDSA_SHA1:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-160)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA256:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-256)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA384:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-384)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA512:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-512)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+
+      case RSA_PSS_SHA256:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-256"));
+      case RSA_PSS_SHA384:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-384"));
+      case RSA_PSS_SHA512:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-512"));
+
+      // for DSA, ECDSA, GOST parameters "SHALL" be empty
+      case DSA_SHA1:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-160)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA256:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-256)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA384:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-384)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA512:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-512)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+
+      default:
+         return AlgorithmIdentifier();
+      }
```

Comment:
This is somewhat a duplicate of the implementations of `EMSA::config_for_x509()` scattered around the different EMSAs. I'm not quite happy with that, to be honest. Simply instantiating the respective EMSA and then actually using `::config_for_x509()` doesn't work out-of-the-box, as it would currently require to pass a respective private key. I think it would be helpful to abstract both use cases and centralize this somehow.

Anyway: As is, this is fairly inefficient, as the `AlgorithmIdentifier`s are re-generated for each invocation. It would be an interesting exercise to `constexpr` the heck out of this. 🤡

I'm open for better ideas. :)

Reply:
Yeah pretty ugly. Also we recompute the PSS params each time when in practice it will always be SHA-{256,512} :sob: 

Looking at `config_for_x509` ... why the heck does this interface take a `Private_Key`? It could easily get away with a `Public_Key`? (Not that this change would address the issue here, but still kind of odd).

Anyway this falls in the category of definitely not great and pointing to a place where some refactoring time is needed but doesn't IMO need to block the PR merging.

Reply:
After merging #3142, this now became: https://github.com/randombit/botan/pull/3125#discussion_r1070925009
Thanks for the help in adapting `config_for_x509()` to make it suitable for this use case!